### ISSUE TITLE
Some Parker structure models result in unphysical velocities/densities

### ISSUE THREAD

**Issue by ladsantos** [ladsantos/p-winds#31]
I was just trying to make `parker.structure()` models for HD 189733 b, and I am getting unphysical values for the velocities and densities, which in turn result in errors when calculating the `hydrogen.ion_fraction()`.

Steps to reproduce the bug:
```
import numpy as np
import astropy.units as u
from p_winds import hydrogen, tools


# HD 189733 b
R_pl = 1.13  # Planetary radius (Jupiter radii)
M_pl = 1.13  # Planetary mass (Jupiter masses)
a_pl = 0.031  # Semi-major axis (au)
planet_to_star_ratio = 0.1504
impact_parameter = 0.656
h_fraction = 0.90  # H number fraction
he_fraction = 1 - h_fraction  # He number fraction
he_h_fraction = he_fraction / h_fraction
mean_f_ion = 0.90  # Initially assumed, but the model relaxes for it
mu_0 = (1 + 4 * he_h_fraction) / (1 + he_h_fraction + mean_f_ion)
initial_f_ion = 0.0  # Fraction of ionized hydrogen

log_m_dot_0 = np.log10(1E9)  # Planetary mass loss rate (g / s)
log_T_0 = np.log10(9000)  # Atmospheric temperature (K)
v_wind_0 = -2E3  # Line-of-sight wind velocity (m / s)

m_dot = 10 ** log_m_dot_0
T = 10 ** log_T_0

r = np.logspace(0, np.log10(20), 100)

relax_solution = True  # This will iteratively relax the solutions until convergence
exact_phi = True  # Exact calculation of H photoionization

host_spectrum = tools.generate_muscles_spectrum('v-eps-eri',
                                                '../data/MUSCLES/',  # Path to MUSCLES spectra
                                                 semi_major_axis=8.84)

f_r, mu_bar = hydrogen.ion_fraction(r, R_pl, T, h_fraction,
                                    m_dot, M_pl, mu_0,
                                    spectrum_at_planet=host_spectrum,
                                    initial_f_ion=initial_f_ion,
                                    relax_solution=relax_solution,
                                    exact_phi=exact_phi, return_mu=True)
```
This code yields the following error:
```
RuntimeError                              Traceback (most recent call last)
/var/folders/0q/ph_vqx8j58s5t8b4smvzpcf80002vw/T/ipykernel_79386/3009913113.py in <module>
      2 r = np.logspace(0, np.log10(20), 100)  # Radial distance profile in unit of planetary radii
      3 
----> 4 f_r, mu_bar = hydrogen.ion_fraction(r, R_pl, T_0, h_fraction, 
      5                             m_dot_0, M_pl, mu_0,
      6                             spectrum_at_planet=spectrum, exact_phi=True,

~/Coding/p-winds/p_winds/hydrogen.py in ion_fraction(radius_profile, planet_radius, temperature, h_fraction, mass_loss_rate, planet_mass, mean_molecular_weight_0, spectrum_at_planet, flux_euv, initial_f_ion, relax_solution, convergence, max_n_relax, exact_phi, return_mu, **options_solve_ivp)
    465             # of `r`
    466             if len(f_r) != len(r):
--> 467                 raise RuntimeError('The solver ``solve_ivp`` failed to obtain a'
    468                                    ' solution.')
    469 

RuntimeError: The solver ``solve_ivp`` failed to obtain a solution.
```

**Comment by ladsantos** [comment]
The issue seems to arise from the calculation of the following transcendental equation:
<img width="289" alt="Screen Shot 2022-02-04 at 11 14 13 AM" src="https://user-images.githubusercontent.com/2818188/152563487-59b2e8ef-4725-42fe-bf2c-3d8ee6834d50.png">
between lines 184 and 209 of the `parker` module. In the case of HD 189733 b, it is yielding negative values of velocity, which are the source of the problem.

**Comment by ladsantos** [comment]
On further digging, it seems to me that this may be a problem that is difficult to solve due to the stiffness of this numerical setup. The best workaround I found was to avoid computing in the region closest to the planetary surface, where the density gradient explodes. Models with very compact upper atmospheres, such as HD 189733 b, will likely always suffer from this issue in this calculation.

**Comment by ladsantos** [comment]
This issue has mostly been resolved in versions >1.4
